src/media_streams/mod.rs

```diff
@@ -0,0 +1,215 @@
+//! Primitives of the Media Capture and Streams API
+//!
+//! The Media Capture and Streams API, often called the Media Streams API or MediaStream API, is an
+//! API related to WebRTC which provides support for streaming audio and video data.
+//!
+//! It provides the interfaces and methods for working with the streams and their constituent
+//! tracks, the constraints associated with data formats, the success and error callbacks when
+//! using the data asynchronously, and the events that are fired during the process.
+//!
+//! <https://developer.mozilla.org/en-US/docs/Web/API/Media_Capture_and_Streams_API>
+
+use crate::{AudioBuffer, FallibleBuffer};
+use arc_swap::ArcSwap;
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+
+/// Ready-state of a [`MediaStreamTrack`]
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum MediaStreamTrackState {
+    /// The track is active (the track's underlying media source is making a best-effort attempt to
+    /// provide data in real time).
+    Live,
+    /// The track has ended (the track's underlying media source is no longer providing data, and
+    /// will never provide more data for this track). Once a track enters this state, it never
+    /// exits it.
+    Ended,
+}
+
+/// Single media track within a [`MediaStream`]
+#[derive(Clone)]
+pub struct MediaStreamTrack {
+    inner: Arc<MediaStreamTrackInner>,
+}
+
+struct MediaStreamTrackInner {
+    data: ArcSwap<FallibleBuffer>,
+    position: AtomicU64,
+    ended: AtomicBool,
+    provider: Mutex<Box<dyn Iterator<Item = FallibleBuffer> + Send + Sync + 'static>>,
+}
+
+impl MediaStreamTrack {
+    pub fn lazy<T: IntoIterator<Item = FallibleBuffer>>(iter: T) -> Self
```

Comment:
Why `lazy` rather than `new`?

Reply:
I will write some extensive documentation for this module. I'm preparing for two kinds of streams:
- lazy: these have no internal clock and only progress when polled. 
- forced: driven by an external clock mechanism

Streaming audio from a mic or webrtc will be a forced stream. They possibly need buffering to synchronise with the web audio context clock. 
Streaming audio from a disk or generating samples on the fly could benefit from the lazy stream as they will precisely follow the web audio context time progression

To be continued!

Reply:
Maybe better names would be `new` for forced streams and `from_iter` for lazy. I expect lazy streams to be more niche. 

Both are obviously for advanced users only. Producing and consuming streams manually is only necessary when microphone and MediaElement input or speaker output is not sufficient for an application.